### PR TITLE
[ie/KiKA] Add extractor

### PR DESCRIPTION
**Pull request by 1100101**
### Description of your *pull request* and other information

</details>

Add extractor for www.kika.de
The website was previously handled by the `MDR` extractor, but that one is outdated and does not work anymore for kika.de

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/kika.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
```
Comment (pukkandan) [inline yt_dlp/extractor/kika.py:3]: remove

File: yt_dlp/extractor/kika.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    determine_ext,
+    int_or_none,
+    parse_duration,
+    parse_iso8601
+)
+
+
+class KikaIE(InfoExtractor):
+    IE_DESC = 'KiKA.de'
+    _VALID_URL = r'https?://(?:www\.)?kika\.de/(?:.*)/[a-z-]+-?(?P<id>\d+)(?:_.+?)?'
```
Comment (pukkandan) [inline yt_dlp/extractor/kika.py:15]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?kika\.de/[^/?#]+/[\w-]+(?P<id>\d+)'
```

File: yt_dlp/extractor/kika.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            vtt = {
+                'url': webvtt_resource,
+                'ext': 'webvtt'
+            }
+            subtitles['de'].append(vtt)
+
+        return {
+            'id': video_id,
+            'title': title,
+            'description': doc['description'],
```
Comment (pukkandan) [inline yt_dlp/extractor/kika.py:64]: make non-fatal

File: yt_dlp/extractor/kika.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        video_id = self._match_id(url)
+
+        doc = self._download_json("https://www.kika.de/_next-api/proxy/v1/videos/video%s" % (video_id), video_id)
+        title = doc.get('title')
+        timestamp = parse_iso8601(doc.get('date'))
+        duration = parse_duration(doc.get('duration'))
+
+        video_url = doc.get('assets').get('url')
+        video_assets = self._download_json(video_url, video_id)
+        formats = self._extract_formats(video_assets, video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/kika.py:44]: unnecessary variables

File: yt_dlp/extractor/kika.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                    'url': stream_url,
+                    'format_id': 'a%s-%s' % (num, ext),
+                    'width': media.get('frameWidth'),
+                    'height': media.get('frameHeight'),
+                    'filesize': int_or_none(media.get('fileSize')),
+                    'abr': int_or_none(media.get('bitrateAudio')),
+                    'vbr': int_or_none(media.get('bitrateVideo')),
+                }
+                formats.append(f)
+        return formats
```
Comment (pukkandan) [inline yt_dlp/extractor/kika.py:93]:
```suggestion
    def _extract_formats(self, media_info, video_id):
        for media in media_info['assets']:
            stream_url = media.get('url')
            if not stream_url:
                continue
            ext = determine_ext(stream_url)
            if ext == 'm3u8':
                yield from self._extract_m3u8_formats(
                    stream_url, video_id, 'mp4', m3u8_id='hls', fatal=False))
            else:
                yield {
                    'url': stream_url,
                    'format_id': ext,
                    **traverse_obj(media, {
                        'width': 'frameWidth',
                        'height': 'frameHeight',
                        'filesize': 'fileSize',
                        'abr': 'bitrateAudio',
                        'vbr': 'bitrateVideo',
                    }, expected_type=int_on_none)
                }
```
cast into list when calling

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/mdr.py [hunk 6]
```diff
@@ -33,30 +33,6 @@ class MDRIE(InfoExtractor):
...
-        'info_dict': {
-            'id': '8182',
-            'ext': 'mp4',
-            'title': 'Beutolomäus und der geheime Weihnachtswunsch',
-            'description': 'md5:b69d32d7b2c55cbe86945ab309d39bbd',
-            'timestamp': 1482541200,
-            'upload_date': '20161224',
-            'duration': 4628,
-            'uploader': 'KIKA',
-        },
```
[the remark this answers is not in the document]
Comment (1100101) [reply yt_dlp/extractor/mdr.py:59]: That video does not exist anymore

File: yt_dlp/extractor/mdr.py [hunk 7]
```diff
@@ -67,28 +43,7 @@ class MDRIE(InfoExtractor):
...
-        'params': {
-            'skip_download': True,
-        },
-    }, {
-        'url': 'http://www.kika.de/baumhaus/sendungen/video19636_zc-fea7f8a0_zs-4bf89c60.html',
-        'only_matching': True,
-    }, {
-        'url': 'http://www.kika.de/sendungen/einzelsendungen/weihnachtsprogramm/einzelsendung2534.html',
-        'only_matching': True,
```
[the remark this answers is not in the document]
Comment (1100101) [reply yt_dlp/extractor/mdr.py:91]: Same here. These videos do not exist anymore.

File: yt_dlp/extractor/kika.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        return {
+            'id': video_id,
+            'title': title,
+            'description': doc['description'],
+            'timestamp': timestamp,
+            'duration': duration,
+            'formats': formats,
+            'subtitles': subtitles,
+            'uploader': 'KIKA'
+        }
```
[the remark this answers is not in the document]
Comment (1100101) [reply yt_dlp/extractor/kika.py]: Not sure this applies here: The framework behind the website has completely changes, the IDs are not the same anymore.
I even suspect the `MDR` extractor does not work anymore.

**Comment by dirkf** [comment]
Call `_self._sort_formats(formats)` after assigning `formats`?

**Comment by bashonly** [comment]
> Call `self._sort_formats(formats)` after assigning `formats`?

no longer needed after 9f14daf22b4080ae1531a772ee7574959af4e2fa

**Comment by pukkandan** [comment]
`_sort_formats` is deprecated amd the functionality moved into core code

**Comment by dirkf** [comment]
In that case `video_assets` may not be a `dict` when used to extract subtitles, if anyone cares.

No doubt the use of the deprecated method as a guard against returning an empty or invalid info_dict in various extractors was taken into account?

**Comment by dirkf** [comment]
This video fails: https://www.kika.de/echt/videos/echt-staffel-drei-darum-geht-s-100.

See https://github.com/ytdl-org/youtube-dl/issues/31428#issuecomment-1355878487. TLDR; the video seems to be staged via ZDF rather than using the Kika JSON (happy to be proved wrong).

**Comment by dirkf** [comment]
Testing from UK:
* HLS manifest needs XFF to fetch, or 403
* HLS formats need XFF to download, or 403
* MP4 manifest doesn't need XFF to fetch
* MP4 formats give 403 for download regardless of XFF.

Probably downloads can't be tested in CI.

**Comment by dirkf** [comment]
Solution:
```diff
...
-    _VALID_URL = r'https?://(?:www\.)?kika\.de/(?:.*)/[a-z-]+-?(?P<id>\d+)'
+    _VALID_URL = r'https?://(?:www\.)?kika\.de/(?:.*)/(?P<id>[a-z-]+-?\d+)'
...
-        doc = self._download_json("https://www.kika.de/_next-api/proxy/v1/videos/video%s" % (video_id), video_id)
+        doc = self._download_json('https://www.kika.de/_next-api/proxy/v1/videos/' + video_id, video_id)
...
```
And the test's ID becomes 'video59362'.

The [series that the yt-dl user wanted](https://github.com/yt-dlp/yt-dlp/pull/5788#issuecomment-1356023194) can be accessed with `--no-geo-bypass`, both manifests and media. Sample hosts are zdfvod-rwrtr.akamaized.net for HLS and tvdlzdf-a.akamaihd.net for mp4: the media links look similar if not identical to those returned using the ZDF extractor for the matching shows; I guess this reflects a different geo-restriction policy of ZDF (no XFF needed) vs Kika. Eg, these shows are the same:
```
https://www.zdf.de/kinder/echt/folge-10-s03-together-forever-100.html
https://www.kika.de/echt/videos/zehn-together-forever-100
```

File: yt_dlp/extractor/mdr.py [hunk 9]
```diff
@@ -67,28 +43,7 @@ class MDRIE(InfoExtractor):
...
-        'url': 'https://www.kika.de/filme/sendung128372_zc-572e3f45_zs-1d9fb70e.html',
-        'info_dict': {
-            'id': '128372',
-            'ext': 'mp4',
-            'title': 'Der kleine Wichtel kehrt zurück',
-            'description': 'md5:f77fafdff90f7aa1e9dca14f662c052a',
-            'duration': 4876,
-            'timestamp': 1607823300,
-            'upload_date': '20201213',
-            'uploader': 'ZDF',
```
Comment (pukkandan) [inline yt_dlp/extractor/mdr.py:81]: This test has uploader != kika. So why are we statically setting it in new code?

File: yt_dlp/extractor/kika.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+            })
+
+        return {
+            'id': video_id,
+            'title': doc.get('title'),
+            'description': doc.get('description'),
+            'timestamp': parse_iso8601(doc.get('date')),
+            'duration': parse_duration(doc.get('duration')),
+            'formats': formats,
+            'subtitles': subtitles,
```
Comment (pukkandan) [inline yt_dlp/extractor/kika.py:58]:
```suggestion
        doc = self._download_json(f'https://www.kika.de/_next-api/proxy/v1/videos/video{video_id}', video_id)
        video_assets = self._download_json(doc['assets']['url'], video_id)

        subtitles = {}
        ttml_resource = video_assets.get('videoSubtitle')
        if ttml_resource:
            subtitles['de'] = [{
                'url': ttml_resource,
                'ext': 'ttml',
            }]
        webvtt_resource = video_assets.get('webvttUrl')
        if webvtt_resource:
            subtitles.setdefault('de', []).append({
                'url': webvtt_resource,
                'ext': 'vtt'
            })

        return {
            'id': video_id,
            'title': doc.get('title'),
            'description': doc.get('description'),
            'timestamp': parse_iso8601(doc.get('date')),
            'duration': parse_duration(doc.get('duration')),
            'formats': list(self._extract_formats(video_assets, video_id)),
            'subtitles': subtitles,
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by pukkandan** [comment]
Also https://github.com/yt-dlp/yt-dlp/pull/5788#issuecomment-1356059090

**Comment by Myzel394** [comment]
Are there any news on this?

**Comment by barkawa** [comment]
Can confirm https://github.com/yt-dlp/yt-dlp/pull/5788#issuecomment-1356059090 works. Tried with `https://www.kika.de/moppi-und-der-leckerladen/videos/sechzehn-fluffis-100`. Do you have the time to incorporate the change @1100101 ?

File: yt_dlp/extractor/mdr.py (hunk 9 above)
Comment (1100101) [reply yt_dlp/extractor/mdr.py:81]: I've removed the static `uploader` tag, it seems that information is not provided anymore by the JSON from the website.

**Comment by tuxflo** [comment]
Are there still changes left?
I really would like to see this merged.

**Comment by Myzel394** [comment]
@tuxflo Most stuff that is on KiKa is also available on ZDF, for which the extractor is working for me, so you can use this as a workaround for now :D

**Comment by tuxflo** [comment]
Yes, there are workarounds. I was also thinking about using Mediathekview as an alternative. But I'm running yt-dlp-bot (https://github.com/tropicoo/yt-dlp-bot) to easily get stuff for the kids downloaded to my NAS by just sending the link in a telegram message. They can also post links to the bot on their own, so if this gets merged I don't have to spend the time to deal with the workarounds 😅

**Review by seproDev — APPROVED** [review]
Hey, truly sorry this took so long. This PR somehow fell off the radar.
I've pushed a slightly reworked version. From my side, this looks good now.

File: yt_dlp/extractor/mdr.py [hunk 11]
```diff
@@ -34,30 +34,6 @@ class MDRIE(InfoExtractor):
             'uploader': 'MITTELDEUTSCHER RUNDFUNK',
         },
         'skip': '404 not found',
-    }, {
-        'url': 'http://www.kika.de/baumhaus/videos/video19636.html',
-        'md5': '4930515e36b06c111213e80d1e4aad0e',
-        'info_dict': {
-            'id': '19636',
```
Comment (bashonly) [inline yt_dlp/extractor/mdr.py:41]: would it make sense to return `_old_archive_ids` in the new extractor?

File: yt_dlp/extractor/kika.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+
+        doc = self._download_json(f'https://www.kika.de/_next-api/proxy/v1/videos/{video_id}', video_id)
+        video_assets = self._download_json(doc['assets']['url'], video_id)
+
+        subtitles = {}
+        if ttml_resource := video_assets.get('videoSubtitle'):
```
Comment (bashonly) [inline yt_dlp/extractor/kika.py:61]:
```suggestion
        if ttml_resource := url_or_none(video_assets.get('videoSubtitle')):
```

File: yt_dlp/extractor/kika.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+            subtitles['de'] = [{
+                'url': ttml_resource,
+                'ext': 'ttml',
+            }]
+        if webvtt_resource := video_assets.get('webvttUrl'):
```
Comment (bashonly) [inline yt_dlp/extractor/kika.py:66]:
```suggestion
        if webvtt_resource := url_or_none(video_assets.get('webvttUrl')):
```

File: yt_dlp/extractor/kika.py [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+                'duration': ((
+                    ('durationInSeconds', {int_or_none}),
+                    ('duration', {parse_duration})), any),
+                'episode_number': ('episodeNumber', {int_or_none}),
+                'season_number': ('season', {int_or_none}),
+            }),
+        }
+
+    def _extract_formats(self, media_info, video_id):
+        for media in traverse_obj(media_info, ('assets', url_or_none(lambda _, v: v['url']))):
```
Comment (bashonly) [inline yt_dlp/extractor/kika.py:90]:
```suggestion
        for media in traverse_obj(media_info, ('assets', lambda _, v: url_or_none(v['url']))):
```

File: yt_dlp/extractor/kika.py [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+                yield from self._extract_m3u8_formats(
+                    stream_url, video_id, 'mp4', m3u8_id='hls', fatal=False)
+            else:
+                yield {
+                    'url': stream_url,
+                    'format_id': ext,
+                    **traverse_obj(media, {
+                        'width': ('frameWidth', {int_or_none}),
+                        'height': ('frameHeight', {int_or_none}),
+                        'filesize': ('fileSize', {int_or_none}, {lambda x: x or None}),
```
Comment (bashonly) [inline yt_dlp/extractor/kika.py:103]: I'm guessing `fileSize` can (incorrectly) be `0` sometimes? might be worth a comment

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/kika.py (hunk 15 above)
Comment (seproDev) [reply yt_dlp/extractor/kika.py:104]: Yes. I've added a comment.

**Review by bashonly — APPROVED** [review]

**Pull request merged**